src/OrleansServiceBus/Providers/Streams/EventHub/EventHubCacheEvictionStrategy.cs

```diff
@@ -0,0 +1,142 @@
+        //buffers which are still in use for current cache
+        /// <summary>
+        /// Buffers which are currently in use in the cache
+        /// </summary>
+        protected readonly Queue<FixedSizeBuffer> inUseBuffers;
+        private readonly TimePurgePredicate timePurge;
+        /// <summary>
+        /// Buffers which are purged
+        /// </summary>
+        protected readonly Queue<FixedSizeBuffer> purgedBuffers;
+        private readonly Logger logger;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="logger"></param>
+        /// <param name="timePurage"></param>
+        public EventHubCacheEvictionStrategy(Logger logger, TimePurgePredicate timePurage = null)
+        {
+            this.inUseBuffers = new Queue<FixedSizeBuffer>();
+            this.logger = logger.GetSubLogger(this.GetType().Name);
+            this.purgedBuffers = new Queue<FixedSizeBuffer>();
+            this.timePurge = timePurage ?? TimePurgePredicate.Default;
+        }
+        /// <inheritdoc />
+        public IPurgeObservable<CachedEventHubMessage> PurgeObservable { private get; set; }
+
+        /// <summary>
+        /// Called with the newest item in the cache and last item purged after a cache purge has run.
+        /// For ordered reliable queues we shouldn't need to notify on every purged event, only on the last event 
+        ///   of every set of events that get purged.
+        /// </summary>
+        public Action<CachedEventHubMessage?, CachedEventHubMessage?> OnPurged { get; set; }
+
+        /// <inheritdoc />
+        public void OnBlockAllocated(IDisposable newBlock)
+        {
+            var newBuffer = newBlock as FixedSizeBuffer;
+            this.inUseBuffers.Enqueue(newBuffer);
+            newBuffer.SetPurgeAction(this.OnFreeBlockRequest);
+        }
+
+        /// <inheritdoc />
+        public void PerformPurge(DateTime nowUtc, IDisposable purgeRequest)
+        {
+            //if the cache is empty, then nothing to purge, return
+            if (this.PurgeObservable.IsEmpty)
+                return;
+            var itemCountBeforePurge = this.PurgeObservable.ItemCount;
+            int itemsPurged = 0;
+            CachedEventHubMessage neweswtMessageInCache = this.PurgeObservable.Newest.Value;
+            CachedEventHubMessage? lastMessagePurged = null;
+            while (!this.PurgeObservable.IsEmpty)
+            {
+                var oldestMessageInCache = this.PurgeObservable.Oldest.Value;
+                if (!ShouldPurge(ref oldestMessageInCache, ref neweswtMessageInCache, nowUtc))
+                {
+                    break;
+                }
+                lastMessagePurged = oldestMessageInCache;
+                itemsPurged++;
+                this.PurgeObservable.RemoveOldestMessage();
+            }
+            var itemCountAfterPurge = itemCountBeforePurge - itemsPurged;
+
+            //purge finished, time to conduct follow up actions 
```

Comment:
these 'follow up' actions should only be performed if itemsPurged != 0, correct?